### DISCUSSION TITLE
[Reactive]: redundant ListState.Value overload?

### DISCUSSION THREAD

**Discussion by weitzhandler** [unoplatform/uno.extensions#1349] [General]
`ListState` offers several factory methods for creating `IListStates<T>` (as well as `ListFeed`), some of which have duplicate overloads for both `Func<IImmutableList<T>>` (interface) and `Func<ImmutableList<T>>` (concrete), for example:

https://github.com/unoplatform/uno.extensions/blob/68b1506584bceb1f11ad7b7ab989c38fbb438c6f/src/Uno.Extensions.Reactive/Core/ListState.cs#L30-L54

What's the reson they both exist, can't we just have only the interface version?  
In fact they could all be named `Create` as their overloads defer.

There are other inconsistensies between the 4 factory classes, for example some don't provide a `Func<CancellationToken, ...` overload where some do only with one. Can you please explain?

Another one - there are inconsistencies in `ListState`'s extensions, for instance `Update` and `UpdateAsync` can be both named just `Update` as they're different overloads. Also `AddAsync` and others.

**Comment by dr1rrb** [comment]
> What's the reson they both exist, can't we just have only the interface version?

The `T` on `IImmutableList` contravariant (nor covariant BTW), so without extension dedicated to the common concrete `ImmutableList<T>` type, user would have to specify the type of `T`.

> In fact they could all be named Create as their overloads defer.

Yes but as lot of theme have delegate as parameter, reducing the number of possible overloads greatly helps type inference for auto-completion.

> There are other inconsistensies between the 4 factory classes, for example some don't provide a `Func<CancellationToken,` ... overload

The overloads that has a `CancellationToken` are only for those which returns a `IAsyncEnumerable` to be used with the [`EnumeratorCancellation`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.enumeratorcancellationattribute?view=net-8.0) so we can use method-group as parameter (And BTW I noticed that one is missing, we should fix it).

> instance Update and UpdateAsync can be both named just Update as they're different overloads

Same, reducing the number of overloads to help type-inference and make it clearer to the user what he has to do.

**Reply by weitzhandler** [reply]
Thanks for your detailed replies @dr1rrb!

So only left with two issues:

- There's one method missing cancellation.
- Some (IListState) some methods have `Async` suffix while others don't.
